### ISSUE TITLE
debounceTime: why returns only last value in any case?

### ISSUE THREAD

**Issue by ValeriusGC** [ReactiveX/rxdart#564]
Hi! According [marble](lib/src/transformers/backpressure/debounce.dart:75) and yours description (if i get it right)
>> This time span start after the last debounced event was emitted.

so i wait something like that (after every time span 200 msec)

`10, 24` etc

but i always get only last value.

Here's the code from your tests, slightly changed

```
Stream<int> _getStream() {
  final controller = StreamController<int>();

  Timer(const Duration(milliseconds: 100), () {
    controller.add(1);
    controller.add(2);
    controller.add(3);
    controller.add(4);
    controller.add(5);
  });
  Timer(const Duration(milliseconds: 200), () {
    controller.add(6);
    controller.add(7);
    controller.add(8);
    controller.add(9);
    controller.add(10);
  });
  Timer(const Duration(milliseconds: 300), () {
    controller.add(11);
    controller.add(12);
    controller.add(13);
    controller.add(14);
    controller.add(15);
  });
  Timer(const Duration(milliseconds: 400), () => controller.add(24));
  Timer(const Duration(milliseconds: 500), () => controller.add(25));
  Timer(const Duration(milliseconds: 600), () => controller.add(26));
  Timer(const Duration(milliseconds: 700), () => controller.add(27));
  Timer(const Duration(milliseconds: 800), () => controller.add(28));
  Timer(const Duration(milliseconds: 900), () => controller.add(29));
  Timer(const Duration(milliseconds: 1000), () => controller.add(210));
  Timer(const Duration(milliseconds: 1100), () {
    controller.add(211);
    controller.close();
  });

  return controller.stream;
}

====================

    _getStream()
        .debounceTime(const Duration(milliseconds: 100))
        .listen((event) {
      print('${DateTime.now().millisecondsSinceEpoch} debounce: $event');
    });

```

**Comment by hoc081098** [comment]
I think that is expected behavior, when an event emitted when previous event still be pending, previous timer will be cancelled.
```
*----*----*-*-*-
---*----*---------*
```

**Comment by ValeriusGC** [comment]
> I think that is expected behavior, when an event emitted when previous event still be pending, previous timer will be cancelled.
> 
> ```
> *----*----*-*-*-
> ---*----*--------*
> ```

I don't think so. Look at this example, where i use generator and two listeners.

```
Stream<int> _get() async* {
  for(var i = 0; i < 15; ++i){
    await delayMilli(100);
    yield i;
  }
}
=============
    _get().debounceTime(const Duration(milliseconds: 200))
        .listen((event) {
      print('$now: debounce: $event');
    });
    _get()
        .listen((event) {
      print('$now: direct: $event');
    });
    await delayMilli(5000);

```
print is like that:
```
2021-03-22 07:37:31.949295: direct: 0
2021-03-22 07:37:32.053612: direct: 1
2021-03-22 07:37:32.154548: direct: 2
2021-03-22 07:37:32.255562: direct: 3
2021-03-22 07:37:32.356699: direct: 4
2021-03-22 07:37:32.458464: direct: 5
2021-03-22 07:37:32.559691: direct: 6
2021-03-22 07:37:32.660568: direct: 7
2021-03-22 07:37:32.761534: direct: 8
2021-03-22 07:37:32.862568: direct: 9
2021-03-22 07:37:32.963530: direct: 10
2021-03-22 07:37:33.064533: direct: 11
2021-03-22 07:37:33.165527: direct: 12
2021-03-22 07:37:33.266550: direct: 13
2021-03-22 07:37:33.369699: debounce: 14
2021-03-22 07:37:33.371797: direct: 14
```
You can see that generator works well, but `debounceTime` catches only last value again.

**Comment by frankpepermans** [comment]
Just to clarify, what is your expected outcome in that last example?

The debounce time starts _after_ an event submits, and the window is 200ms, the async handler emits every 100ms, so the debounce interval effectively resets every 100ms, without emitting anything, that is as expected.

**Comment by hoc081098** [comment]
@ValeriusGC 
![image](https://user-images.githubusercontent.com/36917223/111940857-5956fa80-8b02-11eb-8481-4204071f1e48.png)
That's expected behavior

**Comment by ValeriusGC** [comment]
> Just to clarify, what is your expected outcome in that last example?

Ok, will do explain.
I donno where from yours diagram, but from rxdart source diagram is like that:

![Screenshot from 2021-03-22 11-49-15](https://user-images.githubusercontent.com/3320861/111950871-b05dcf00-8b04-11eb-84e2-af6d68b677a7.png)

let's go there and see

![Screenshot from 2021-03-22 11-50-08](https://user-images.githubusercontent.com/3320861/111950928-c8cde980-8b04-11eb-8d7f-eef0828446de.png)
--
I suppose, that `debounceTime` with window 200 ms swallows all between and exposes rest:

- swallows 1, 3, 5, etc, as they between 200 ms.
- exposes 2,4,6, etc, as they matche window size.

In other words i'd like to apply `debounceTime` to search process, when user enter something like `Washington`, and listener do not run every char but swallows some between them and generates chunks like `wash`, `washing` and then `washington`. That's what my expectations.

**Comment by hoc081098** [comment]
try using `throttleTime(duration, leading: true, trailing: true)`

**Comment by ValeriusGC** [comment]
Finally i got it )))
```
/// Emulates manual printing with pauses
Stream<String> _getStream() async* {
  yield 'h';
  yield 'he';
  yield 'hel';
  await delayMilli(100);
  yield 'hello';
  await delayMilli(300);
  yield 'hello, ';
  yield 'hello, world!';
  await delayMilli(300);
}

void main() {
  test('Rx.debounceTime for search: pauses rule!', () async {
    print('${testDelimiter()}: Rx.debounceTime for search: pauses rule!');

    await expectLater(
        _getStream().debounceTime(const Duration(milliseconds: 200)),
        emitsInOrder(<dynamic>['hello', 'hello, world!']));
  });

}

```
